src/components/rating/rating.ts

```diff
@@ -0,0 +1,308 @@
+import { html, LitElement } from 'lit';
+import { property, queryAll, state } from 'lit/decorators.js';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { classMap } from 'lit/directives/class-map.js';
+import { styleMap } from 'lit/directives/style-map.js';
+import { watch } from '../common/decorators/watch';
+import { Constructor } from '../common/mixins/constructor';
+import { EventEmitterMixin } from '../common/mixins/event-emitter';
+import { SizableMixin } from '../common/mixins/sizable';
+import { styles } from './rating.material.css';
+import { clamp } from '../common/util';
+
+export interface IgcRatingEventMap {
+  igcChange: CustomEvent<number>;
+  igcHover: CustomEvent<number>;
+}
+
+/**
+ * Rating provides insight regarding others' opinions and experiences,
+ * and can allow the user to submit a rating of their own
+ *
+ * @element igc-rating
+ *
+ * @fires igcChange - Emitted when the value of the control changes.
+ * @fires igcHover - Emitted when hover is enabled and the user mouses over a symbol of the rating.
+ *
+ * @csspart base - The main wrapper which holds all of the rating elements.
+ */
+export default class igcRatingComponent extends SizableMixin(
+  EventEmitterMixin<IgcRatingEventMap, Constructor<LitElement>>(LitElement)
+) {
+  /** @private */
+  public static tagName = 'igc-rating';
+
+  /** @private */
+  public static styles = [styles];
+
+  @queryAll('span[part="rating-symbol"]')
+  protected elements!: NodeListOf<HTMLSpanElement>;
+
+  @state()
+  protected hoverValue = -1;
+
+  @state()
+  protected hoverState = false;
+
+  protected get isInteractive() {
+    return !(this.readonly || this.disabled);
+  }
+
+  /** The maximum value for the rating */
+  @property({ type: Number })
+  public max = 5;
+
+  /** The minimum value change allowed. */
+  @property({ type: Number })
+  public precision = 1;
+
+  /**
+   * The symbol which the rating will display.
+   * It also accepts a callback function which gets the current symbol
+   * index so the symbol can be resolved per position.
+   */
+  @property()
+  public symbol: string | ((index: number) => string) = '⭐';
```

Comment:
I'm fine with that change.
@gmurray81 Will you chime on this?

As for the _emptySymbol_ property this would entail an _emptySymbolFormatter_ as well. 
Additionally this should be
discussed with the design team on the implementation and the looks. In the case of decimal step you would have a selected part of the symbol/not a selected part of the symbol/ an empty symbol or would be selected/part empty/empty?

Reply:
It's best if we avoid properties with union/intersection types on the public api, since I don't have a good way to represent them on the public apu other than using object at the c# level.

It would generally be better if properties with multiple types were multiple properties that interacted in a well defined way.

And for a callback style property, if it is formulated as an event, then it will automatically work for Blazor.

Do note though, that for Blazor server, the developer is unable to respond to events synchronously, so it would be best if we avoided putting critical functionality into an event that needs to be responded to synchronously.

So having something declaratice like emptySymbol/symbol would be much friendlier.

The other benefit of making things declarative is that they become much easier to specify from markup, so it's usually a win/win.

Reply:
@gmurray81 
Does this mean that all component properties that accept a JS callback (e.g. custom rendering) should be transformed to event emitters?

Reply:
@DiyanDimitrov 
After a discussion with the design team, we came to the agreement to not exposed an empty symbol property
as it will look 'broken' when the step property is different from 1

Reply:
ideally when a callback is necessary its best to use an actual event, since then it will automatically be available at the blazor level, @rkaraivanov. If it's a scenario where it needs to be a callback instead, then we can still make it work, but its certainly easier if it's the former.

The more important thing to keep in mind is the with blazor we can't synchronously respond to an event or callback being fired and provide a response before the control returns to javascript. This is because in the Blazor Server model, the .net event handler code will actually be running on the server instead of in WebAssembly, so the logic may not even be on the client, so we are not able to make the js thread wait for the handler to process. So if you design something where the user _must_ provide a response in an event/callback , it will need a different way to operate in blazor.

We do have a fallback method where the user can actually configure a JS handler to be hooked up to an event/callback, instead of providing the handler in .NET, and that will be able to execute on the client. But since we try to avoid the developer needing to write JS for the Blazor product except if forced to, then we should generally try to avoid scenarios where this is required. This has been pretty tricky, because sometimes a callback/event is the easiest and simplest approach. We've had to jump through a lot of hoops to create declarative style approaches to avoid having to rely too much on events/callbacks. But as I said above, this effort isn't really wasted because it means that more power can be expressed declaratively, and maybe even from markup, rather than requiring the user to hook handlers up to the element.

Reply:
@rkaraivanov,
IMO It would look 'broken' only if the icon and empty icon are completely different, which is a developer responsibility. This [sample](https://mui.com/components/rating/#customization) demonstrates rating component with precision, icon and empty icon and doesn't look broken.